src/main/java/jenkins/branch/MultiBranchProjectFactory.java

```diff
@@ -0,0 +1,121 @@
+    
+    @Override public MultiBranchProjectFactoryDescriptor getDescriptor() {
+        return (MultiBranchProjectFactoryDescriptor) super.getDescriptor();
+    }
+
+    // TODO we have no way with the current API of instructing an SCMSource to use a specific SCMSourceCriteria during retrieve; impls hardcode SCMSource.getCriteria, which relies on a parent SCMSourceOwner
+    interface SCMSourceOwnerHack extends SCMSourceOwner {
+        <T> T withSCMSourceCriteria(@Nonnull SCMSource source, @Nonnull SCMSourceCriteria criteria, @Nonnull Callable<T> body) throws Exception;
+    }
+
+    /**
+     * Creates a particular kind of multibranch project insofar as at least one {@link SCMHead} satisfies a probe.
+     */
+    public static abstract class BySCMSourceCriteria extends MultiBranchProjectFactory {
+
+        /**
+         * Defines how to decide whether or not a given repository should host our type of project.
+         * @param source a repository
+         * @return criteria for treating its branches as a match
+         */
+        protected abstract @Nonnull SCMSourceCriteria getSCMSourceCriteria(@Nonnull SCMSource source);
+
+        /**
+         * Creates a project given that there seems to be a match.
+         * @param parent the folder
+         * @param name the project name
+         * @param attributes a set of metadata attributes as added by {@link jenkins.scm.api.SCMSourceObserver.ProjectObserver#addAttribute}
+         * @return a new project suitable for {@link #createProject}
+         */
+        protected abstract @Nonnull MultiBranchProject<?,?> doCreateProject(@Nonnull ItemGroup<?> parent, @Nonnull String name, @Nonnull Map<String,Object> attributes);
+
+        @Override public final MultiBranchProject<?,?> createProject(ItemGroup<?> parent, String name, List<? extends SCMSource> scmSources, @Nonnull Map<String,Object> attributes, final TaskListener listener) throws IOException, InterruptedException {
+            for (final SCMSource scmSource : scmSources) {
+                SCMSourceOwner owner = scmSource.getOwner();
+                if (!(owner instanceof SCMSourceOwnerHack)) {
+                    throw new IOException("cannot influence criteria on " + owner);
+                }
+                boolean empty;
+                try {
+                    empty = ((SCMSourceOwnerHack) owner).withSCMSourceCriteria(scmSource, getSCMSourceCriteria(scmSource), new Callable<Boolean>() {
+                        @Override public Boolean call() throws Exception {
+                            return scmSource.fetch(listener).isEmpty();
+                        }
+                    });
+                // This is ugly but lambdas do not seem to help at all: http://stackoverflow.com/q/18198176/12916
+                } catch (IOException x) {
```

Comment:
It might be a bit ugly, but it is very easy to grok what's you want to happen, which trumps brevity for me any day.